### ISSUE TITLE
Training(theory) Documentation

### ISSUE THREAD

**Issue by uday60** [AlexeyAB/darknet#796]
Is there some documentation available to know what happens during training in particular to yolo so that we can have a deeper understanding ? I am more interested in the theory. 

Thanks.

**Comment by uday60** [comment]
This is going to keep me busy for a while. I will get back to asking further questions. If possible can we link this post the main page so that others will also benefit from the same. 

Thank you.

**Comment by uday60** [comment]
Let me put this in normal English: (please correct me if i am wrong)

1. Suppose there is an Image with A written as text.
2. CNN will try to match parts of the image rather than the whole thing.
3. CNN breaks the image into smaller parts called features.
4. These smaller parts are then matched with the image using filtering. 
 a) Feature(part) of the image is multiplied by each pixel of the corresponding image.(-1 or 1)
 b) Then adding these pixel values(many -1 and 1) and then dividing them with no of pixels.( for eg 12/12 = 1 for perfect match )
5. By moving it around every possible position we get a nice map of the image matching the features. 
6. By repeating the same with other features the map becomes consistent with the features that are matched. 
7. This whole process happens in different layers.
8. Max Pooling
-----------------
Shrinking the image and taking the maximum value(0.12, -0.34, 1 ; we take the max i.e 1)

9. RELU
--------
Changing all the -ve value to 0

10. So there can be many layers like these one after the other.
11. Final layer(many) we rearrange the values(2 D array) in a single list where each value with weights points to some output. The average of the weights decide what the object is most likely to be.

12. Backpropogation
---------------------
If while predicting the output the accuracy is 0.92 there is error of 0.08. Now this error will force the Gradient descent algorithm which will make adjustments in the values. The greater the error the bigger the adjustment. Small error = small adjustment. No error = no adjustment. 

The objective is to reduce the error rate and it keeps on going till the error is the least. 

I will edit this post after going through the Yolo articles.

**Comment by AlexeyAB** [comment]
@uday60  If you keep it simple, then yes.

But there we don't divide it by the number of pixels.

> b) Then adding these pixel values(many -1 and 1) and then dividing them with no of pixels.( for eg 12/12 = 1 for perfect match )
